dev-tools/omdb/src/bin/omdb/nexus.rs

```diff
@@ -1048,3 +1085,109 @@ async fn cmd_nexus_sled_add(
     Ok(())
 }
+
+/// Runs `omdb nexus sleds expunge`
+async fn cmd_nexus_sled_expunge(
+    client: &nexus_client::Client,
+    args: &SledExpungeArgs,
+    omdb: &Omdb,
+    log: &slog::Logger,
+    _destruction_token: DestructiveOperationToken,
+) -> Result<(), anyhow::Error> {
+    // This is an extremely dangerous and irreversible operation. We put a
+    // couple of safeguards in place to ensure this cannot be called without
+    // due consideration:
+    //
+    // 1. We'll require manual input on stdin to confirm the sled to be removed
+    // 2. We'll warn sternly if the sled-to-be-expunged is still present in the
+    //    most recent inventory collection
+    use nexus_db_queries::context::OpContext;
+
+    let datastore = args.db_url_opts.connect(omdb, log).await?;
+    let opctx = OpContext::for_tests(log.clone(), datastore.clone());
+    let opctx = &opctx;
+
+    // First, we need to look up the sled so we know its serial number.
+    let (_authz_sled, sled) = LookupPath::new(opctx, &datastore)
+        .sled_id(args.sled_id)
+        .fetch()
+        .await
+        .with_context(|| format!("failed to find sled {}", args.sled_id))?;
+
+    // Helper to get confirmation messages from the user.
+    let mut line_editor = Reedline::create();
+    let mut read_with_prompt = move |message: &str| {
+        let prompt = DefaultPrompt::new(
+            DefaultPromptSegment::Basic(message.to_string()),
+            DefaultPromptSegment::Empty,
+        );
+        if let Ok(reedline::Signal::Success(input)) =
+            line_editor.read_line(&prompt)
+        {
+            Ok(input)
+        } else {
+            bail!("expungement aborted")
+        }
+    };
+
+    // Now check whether its sled-agent or SP were found in the most recent
+    // inventory collection.
+    match datastore
+        .inventory_get_latest_collection(opctx)
+        .await
+        .context("loading latest collection")?
+    {
+        Some(collection) => {
+            let sled_baseboard = BaseboardId {
+                part_number: sled.part_number().to_string(),
+                serial_number: sled.serial_number().to_string(),
+            };
+            // Check the collection for either the sled-id or the baseboard. In
+            // general a sled-agent's self report should result in both the
+            // sled-id and the baseboard being present in the collection, but
+            // there are some test environments (e.g., `omicron-dev run-all`)
+            // where that isn't guaranteed.
+            let sled_present_in_collection =
+                collection.sled_agents.contains_key(&args.sled_id)
+                    || collection.baseboards.contains(&sled_baseboard);
+            if sled_present_in_collection {
+                eprintln!(
+                    "WARNING: sled {} is PRESENT in the most recent inventory \
+                    collection; are you sure you want to mark it expunged?",
+                    args.sled_id,
```

Comment:
I'd suggest elaborating a bit more about the problem, just to give someone who doesn't know what they're doing a clue about the risk they're taking.  Maybe:

> WARNING: sled {} is PRESENT in the most recent inventory collection (spotted at {timestamp}).  Sleds should be removed prior to being expunged (and not re-inserted after).  It is dangerous to expunge a sled that is still running.  Are you sure you want to proceed anyway?